### PR TITLE
Support for promises and and ES2017 async/await in before/it/after

### PR DESCRIPTION
**Pull request by sgravrock**
This is similar to #1270 but updated to account for recent changes to QueueRunner. I also added support for async functions, since that was nearly free once promises were supported.

### CODE REVIEW

File: spec/helpers/asyncAwait.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+(function(env) {
+  function getAsyncCtor() {
+    try {
+      eval("var func = async function(){};");
```
Comment (sgravrock) [inline spec/helpers/asyncAwait.js]: This is nasty but I haven't been able to find a way to do this that neither requires eval nor breaks in environments that don't support async.

File: src/core/Env.js [hunk 2]
```diff
@@ -319,6 +319,12 @@ getJasmineRequireObj().Env = function(j$) {
       }
     };
 
+    var ensureIsFunctionOrAsync = function(fn, caller) {
```
Comment (sgravrock) [inline src/core/Env.js]: I'm not completely wild about the naming here or in base.js, but I think that most Jasmine code that asks "is this thing a function?" isn't prepared to handle async functions. So it's probably best for j$.isFunction and ensureIsFunction to continue to reject async functions.

**Comment by NicBright** [comment]
This is great. What about documentation? Can't find it :(

**Comment by slackersoft** [comment]
The API docs have a bit about it here: https://jasmine.github.io/api/edge/global.html#implementationCallback. We're still looking into how best to format all of the documentation for Jasmine. The big long pages like introduction.html are pretty and can read nicely, but they need to be broken apart a bit to be more useful.

Hope this helps. Thanks for using Jasmine!

**Pull request merged**